File: src/ai.rs
use std::collections::{HashMap, VecDeque};
use std::sync::{Mutex, OnceLock};

use serde::Deserialize;

use crate::Error;

struct HistoryItem {
    role: String,
    content: String,
}

fn history_map() -> &'static Mutex<HashMap<u64, VecDeque<HistoryItem>>> {
    static HISTORY: OnceLock<Mutex<HashMap<u64, VecDeque<HistoryItem>>>> = OnceLock::new();
    HISTORY.get_or_init(|| Mutex::new(HashMap::new()))
}

fn snapshot(channel: u64) -> Vec<HistoryItem> {
    history_map()
        .lock()
        .unwrap_or_else(|e| e.into_inner())
        .get(&channel)
        .map(|q| {
            q.iter()
                .map(|e| HistoryItem {
                    role: e.role.clone(),
                    content: e.content.clone(),
                })
                .collect()
        })
        .unwrap_or_default()
}

fn push(channel: u64, role: String, content: String) {
    let mut map = history_map().lock().unwrap_or_else(|e| e.into_inner());
    if !map.contains_key(&channel) && map.len() >= 200 {
        if let Some(k) = map.keys().next().cloned() {
            map.remove(&k);
        }
    }
    let q = map.entry(channel).or_insert_with(VecDeque::new);
    q.push_back(HistoryItem { role, content });
    while q.len() > 10 {
        q.pop_front();
    }
    loop {
        let total: usize = q.iter().map(|e| e.content.len()).sum();
        if total <= 3000 || q.is_empty() {
            break;
        }
        q.pop_front();
    }
}

fn client() -> &'static reqwest::Client {
    static CLIENT: OnceLock<reqwest::Client> = OnceLock::new();
    CLIENT.get_or_init(|| {
        reqwest::Client::builder()
            .timeout(std::time::Duration::from_secs(180))
            .build()
            .expect("reqwest client builds")
    })
}

pub(crate) fn default_model() -> String {
    "llama3.1".to_string()
}

pub(crate) fn default_host() -> String {
    "http://127.0.0.1:11434".to_string()
}

pub(crate) fn resolve_host(configured: &str) -> String {
    for key in ["OLLAMA_HOST", "OLLAMA_URL"] {
        if let Ok(v) = std::env::var(key) {
            let v = v.trim().trim_end_matches('/').to_string();
            if !v.is_empty() {
                return v;
            }
        }
    }
    let v = configured.trim().trim_end_matches('/').to_string();
    if v.is_empty() {
        default_host()
    } else {
        v
    }
}

pub(crate) fn valid_model_name(s: &str) -> bool {
    let s = s.trim();
    if s.is_empty() || s.len() > 128 {
        return false;
    }
    if s.starts_with(['.', '-', '/', ':']) || s.ends_with(['.', '-', '/', ':']) {
        return false;
    }
    if s.contains("..") || s.contains("//") || s.contains(' ') {
        return false;
    }
    s.chars()
        .all(|c| c.is_ascii_alphanumeric() || matches!(c, '.' | '_' | '-' | ':' | '/'))
}

#[derive(Deserialize)]
struct ChatResponse {
    #[serde(default)]
    message: Option<ChatMessageOwned>,
    #[serde(default)]
    response: Option<String>,
}

#[derive(Deserialize)]
struct ChatMessageOwned {
    #[serde(default)]
    content: String,
}

pub(crate) fn clean_reply(text: &str) -> String {
    let mut collapsed = String::new();
    let mut blanks = 0;
    for line in text.lines() {
        if line.trim().is_empty() {
            blanks += 1;
            if blanks <= 1 {
                collapsed.push('\n');
            }
        } else {
            blanks = 0;
            collapsed.push_str(line);
            collapsed.push('\n');
        }
    }
    collapsed.trim().to_string()
}

pub(crate) fn strip_leading_speaker(text: &str, names: &[String]) -> String {
    let mut out = text.trim_start().to_string();
    loop {
        let t = out.trim_start();
        if !t.starts_with('[') {
            break;
        }
        let end = match t.find(']') {
            Some(e) if e > 1 && e <= 65 => e,
            _ => break,
        };
        let after = t[end + 1..].trim_start();
        if !after.starts_with(':') {
            break;
        }
        out = after[1..].trim_start().to_string();
    }
    let mut ordered: Vec<&String> = names.iter().collect();
    ordered.sort_by_key(|n| std::cmp::Reverse(n.len()));
    loop {
        let t = out.trim_start();
        let mut hit = false;
        for n in &ordered {
            let n = n.trim();
            if n.is_empty() {
                continue;
            }
            if t.len() > n.len() && t[..n.len()] == *n && t[n.len()..].starts_with(':') {
                out = t[n.len() + 1..].trim_start().to_string();
                hit = true;
                break;
            }
        }
        if !hit {
            break;
        }
    }
    out
}

fn meta_preamble_len(text: &str) -> usize {
    let starters = [
        "let me summarize",
        "just to sum it up",
        "just to summarize",
        "let's simplify",
        "let me simplify",
        "to simplify,",
        "too much info",
        "i got carried away",
        "i think i got",
        "here is a summary",
        "here's a summary",
        "to summarize,",
        "in summary,",
    ];
    let t = text.trim_start();
    let low = t.to_lowercase();
    for s in starters {
        if !low.starts_with(s) {
            continue;
        }
        let rest = &t[s.len()..];
        if s.ends_with(',') || rest.starts_with([':', ',']) {
            let cut = if rest.starts_with([':', ',']) {
                s.len() + 1
            } else {
                s.len()
            };
            if t[cut..].trim().is_empty() {
                return 0;
            }
            return cut;
        }
        let cut = match rest.find(['.', '!', '?']) {
            Some(i) => s.len() + i + 1,
            None => return 0,
        };
        if t[cut..].trim().is_empty() {
            return 0;
        }
        return cut;
    }
    0
}

pub(crate) fn strip_meta_preamble(text: &str) -> String {
    let mut out = text.to_string();
    loop {
        let cut = meta_preamble_len(&out);
        if cut == 0 {
            break;
        }
        out = out.trim_start()[cut..].trim_start().to_string();
    }
    out
}

pub(crate) fn sanitize_reply(raw: &str, names: &[String]) -> String {
    strip_meta_preamble(&strip_leading_speaker(&clean_reply(raw), names))
}

pub(crate) fn finalize_reply(
    channel: u64,
    tagged: String,
    names: &[String],
    raw: &str,
) -> Result<String, Error> {
    let text = sanitize_reply(raw, names);
    if text.trim().is_empty() {
        return Err("ollama returned an empty reply".into());
    }
    push(channel, "user".to_string(), tagged);
    push(channel, "assistant".to_string(), text.clone());
    Ok(text)
}

#[derive(Deserialize)]
struct TagsResponse {
    #[serde(default)]
    models: Vec<TagEntry>,
}

#[derive(Deserialize)]
struct TagEntry {
    #[serde(default)]
    name: String,
}

pub(crate) fn is_rate_limit_err(s: &str) -> bool {
    let t = s.to_lowercase();
    t.contains("429")
        || t.contains("rate limit")
        || t.contains("rate_limit")
        || t.contains("rate-limited")
        || t.contains("too many requests")
        || t.contains("quota")
}

pub(crate) fn is_api_full_err(s: &str) -> bool {
    if is_rate_limit_err(s) {
        return true;
    }
    let t = s.to_lowercase();
    t.contains("503")
        || t.contains("529")
        || t.contains("overload")
        || t.contains("capacity")
        || t.contains("server is busy")
        || t.contains("try again in a bit")
        || t.contains("api full")
}

pub(crate) fn api_full_message() -> String {
    "Sorry, I'm running hot right now (API full/rate limited) — try again in a minute.".to_string()
}

async fn chat_once(
    url: &str,
    model: &str,
    messages: &[serde_json::Value],
) -> Result<ChatMessageOwned, Error> {
    let req = serde_json::json!({
        "model": model,
        "messages": messages,
        "stream": false,
    });
    let resp = client().post(url).json(&req).send().await?;
    if !resp.status().is_success() {
        let status = resp.status();
        let body = resp.text().await.unwrap_or_default();
        let body: String = body.chars().take(300).collect();
        return Err(format!("ollama {status}: {body}").into());
    }
    let parsed: ChatResponse = resp.json().await?;
    if let Some(m) = parsed.message {
        return Ok(m);
    }
    if let Some(r) = parsed.response {
        if !r.trim().is_empty() {
            return Ok(ChatMessageOwned { content: r });
        }
    }
    Err("ollama returned an empty reply".into())
}

pub(crate) fn stale_history_line(s: &str) -> bool {
    let t = s.trim().to_lowercase();
    t.contains("running hot right now")
}

pub(crate) async fn glitch_text(host: &str, model: &str, system_prompt: &str) -> String {
    // Don't hammer a full API with another call — return a helpful static fallback.
    // Callers that already know the error was api-full should prefer api_full_message().
    let url = format!("{}/api/chat", host.trim_end_matches('/'));
    let mut messages = Vec::new();
    if !system_prompt.trim().is_empty() {
        messages.push(serde_json::json!({"role": "system", "content": system_prompt}));
    }
    messages.push(
        serde_json::json!({"role": "user", "content": "You just glitched out. Tell the user in one short sentence, no details."}),
    );
    match chat_once(&url, model, &messages).await {
        Ok(m) => {
            let text = strip_meta_preamble(&strip_leading_speaker(&clean_reply(&m.content), &[]));
            if text.trim().is_empty() {
                "sorry, glitched out — try again in a sec".to_string()
            } else if is_api_full_err(&text) {
                api_full_message()
            } else {
                text.chars().take(300).collect()
            }
        }
        Err(e) if is_api_full_err(&e.to_string()) => api_full_message(),
        Err(_) => "sorry, glitched out — try again in a sec".to_string(),
    }
}

pub(crate) async fn ollama_chat(
    host: &str,
    model: &str,
    channel: u64,
    speaker: &str,
    prompt: &str,
    system_prompt: &str,
) -> Result<String, Error> {
    let host = host.trim_end_matches('/');
    let url = format!("{host}/api/chat");
    let tagged = format!("[{}]: {}", speaker_tag(speaker), prompt);
    let past = snapshot(channel);
    let mut names: Vec<String> = vec![speaker_tag(speaker)];
    for e in &past {
        if e.role == "assistant" {
            continue;
        }
        let t = e.content.trim_start();
        if let Some(rest) = t.strip_prefix('[') {
            if let Some(end) = rest.find(']') {
                if end > 0 && end <= 64 {
                    let n = rest[..end].trim().to_string();
                    if !n.is_empty() && !names.contains(&n) {
                        names.push(n);
                    }
                }
            }
        }
    }
    let mut messages: Vec<serde_json::Value> = Vec::with_capacity(past.len() + 2);
    if !system_prompt.trim().is_empty() {
        messages.push(serde_json::json!({"role": "system", "content": system_prompt}));
    }
    for e in &past {
        let role = if e.role == "assistant" {
            "assistant"
        } else {
            "user"
        };
        if role == "assistant" {
            let cleaned = sanitize_reply(&e.content, &names);
            if cleaned.trim().is_empty() || stale_history_line(&cleaned) {
                continue;
            }
            messages.push(serde_json::json!({"role": role, "content": cleaned}));
        } else {
            messages.push(serde_json::json!({"role": role, "content": e.content}));
        }
    }
    messages.push(serde_json::json!({"role": "user", "content": tagged.clone()}));
    // Repeat the backstory last so the model follows the current TOML
    // prompt strictly instead of drifting into old history style.
    if !system_prompt.trim().is_empty() {
        messages.push(serde_json::json!({"role": "system", "content": system_prompt}));
    }
    let first = match chat_once(&url, model, &messages).await {
        Ok(m) => m,
        Err(e) if is_api_full_err(&e.to_string()) => {
            eprintln!("ollama_chat: api full on first call, offline fallback");
            let fb = api_full_message();
            push(channel, tagged.clone(), tagged.clone());
            push(channel, "assistant".to_string(), fb.clone());
            return Ok(fb);
        }
        Err(e) => return Err(e),
    };
    finalize_reply(channel, tagged, &names, &first.content)
}

pub(crate) fn clear_history(channel: u64) {
    if let Ok(mut map) = history_map().lock() {
        map.remove(&channel);
    }
}

pub(crate) fn clear_all_history() {
    if let Ok(mut map) = history_map().lock() {
        map.clear();
    }
}

pub(crate) fn record_artixy(channel: u64, text: &str) {
    let t = text.trim();
    if t.is_empty() {
        return;
    }
    let kept: String = t.chars().take(1500).collect();
    push(channel, "assistant".to_string(), kept);
}

pub(crate) async fn model_present(host: &str, model: &str) -> Option<bool> {
    let url = format!("{}/api/tags", host.trim_end_matches('/'));
    let resp = client().get(&url).send().await.ok()?;
    if !resp.status().is_success() {
        return None;
    }
    let tags: TagsResponse = resp.json().await.ok()?;
    let want = model.trim().to_lowercase();
    let base = want.split(':').next().unwrap_or(&want);
    Some(tags.models.iter().any(|m| {
        let n = m.name.to_lowercase();
        n == want || n.split(':').next().unwrap_or(&n) == base
    }))
}

pub(crate) fn speaker_tag(raw: &str) -> String {
    let flat: String = raw.split_whitespace().collect::<Vec<_>>().join(" ");
    let clean: String = flat
        .chars()
        .map(|c| match c {
            '[' | ']' => ' ',
            c if c.is_control() => ' ',
            c => c,
        })
        .collect::<String>()
        .split_whitespace()
        .collect::<Vec<_>>()
        .join(" ");
    let clean: String = clean.chars().take(64).collect();
    if clean.trim().is_empty() {
        "someone".to_string()
    } else {
        clean
    }
}

pub(crate) fn strip_mention(content: &str, bot_id: u64) -> String {
    content
        .replace(&format!("<@{bot_id}>"), "")
        .replace(&format!("<@!{bot_id}>"), "")
        .trim()
        .to_string()
}

pub(crate) fn mentions_name(content: &str) -> bool {
    content.to_lowercase().contains("artixy")
}

pub(crate) fn strip_name(content: &str) -> String {
    content
        .split_whitespace()
        .filter(|w| !w.to_lowercase().contains("artixy"))
        .collect::<Vec<_>>()
        .join(" ")
}

pub(crate) fn chunk_reply(s: &str) -> Vec<String> {
    const MAX: usize = 1900;
    const MAX_CHUNKS: usize = 4;
    let s = s.trim();
    if s.chars().count() <= MAX {
        return vec![s.to_string()];
    }
    let mut chunks = Vec::new();
    let mut cur = String::new();
    let mut cur_len = 0usize;
    for line in s.lines() {
        let line_len = line.chars().count() + 1;
        if line_len > MAX {
            if !cur.trim().is_empty() {
                chunks.push(cur.trim_end().to_string());
                cur = String::new();
                cur_len = 0;
            }
            let chars: Vec<char> = line.chars().collect();
            for piece in chars.chunks(MAX) {
                chunks.push(piece.iter().collect());
                if chunks.len() >= MAX_CHUNKS {
                    break;
                }
            }
            if chunks.len() >= MAX_CHUNKS {
                break;
            }
            continue;
        }
        if cur_len + line_len > MAX {
            chunks.push(cur.trim_end().to_string());
            cur = String::new();
            cur_len = 0;
            if chunks.len() >= MAX_CHUNKS {
                break;
            }
        }
        cur.push_str(line);
        cur.push('\n');
        cur_len += line_len;
    }
    if !cur.trim().is_empty() && chunks.len() < MAX_CHUNKS {
        chunks.push(cur.trim_end().to_string());
    }
    if chunks.is_empty() {
        vec![s.chars().take(MAX).collect()]
    } else {
        chunks
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn rate_limit_detects_429_and_quota() {
        assert!(is_rate_limit_err("ollama 429: too many requests"));
        assert!(is_rate_limit_err("Rate limit exceeded, try again"));
        assert!(is_rate_limit_err("quota exceeded"));
        assert!(!is_rate_limit_err("connection refused"));
    }

    #[test]
    fn api_full_covers_overload_and_503() {
        assert!(is_api_full_err("ollama 503: overloaded"));
        assert!(is_api_full_err("server is busy, try again in a bit"));
        assert!(is_api_full_err("ollama 429: too many requests"));
        assert!(!is_api_full_err("connection refused"));
    }

    #[test]
    fn stale_history_skips_api_full_lines() {
        assert!(stale_history_line(&api_full_message()));
    }
}

File: internal/ai/ai.go
package ai

import (
	"bytes"
	"container/list"
	"encoding/json"
	"fmt"
	"io"
	"net/http"
	"os"
	"strings"
	"sync"
	"time"
)

type HistoryItem struct {
	Role    string
	Content string
}

var (
	histMu sync.Mutex
	hist   = map[uint64]*list.List{}
)

func snapshot(channel uint64) []HistoryItem {
	histMu.Lock()
	defer histMu.Unlock()
	l, ok := hist[channel]
	if !ok {
		return nil
	}
	var out []HistoryItem
	for e := l.Front(); e != nil; e = e.Next() {
		out = append(out, e.Value.(HistoryItem))
	}
	// copy
	cp := make([]HistoryItem, len(out))
	copy(cp, out)
	return cp
}

func push(channel uint64, role, content string) {
	histMu.Lock()
	defer histMu.Unlock()
	if _, ok := hist[channel]; !ok {
		if len(hist) >= 200 {
			for k := range hist {
				delete(hist, k)
				break
			}
		}
		hist[channel] = list.New()
	}
	l := hist[channel]
	l.PushBack(HistoryItem{Role: role, Content: content})
	for l.Len() > 10 {
		l.Remove(l.Front())
	}
	for {
		total := 0
		for e := l.Front(); e != nil; e = e.Next() {
			total += len(e.Value.(HistoryItem).Content)
		}
		if total <= 3000 || l.Len() == 0 {
			break
		}
		l.Remove(l.Front())
	}
}

var httpClient = &http.Client{Timeout: 180 * time.Second}

func DefaultModel() string { return "llama3.1" }
func DefaultHost() string  { return "http://127.0.0.1:11434" }

func ResolveHost(configured string) string {
	for _, key := range []string{"OLLAMA_HOST", "OLLAMA_URL"} {
		if v := strings.TrimRight(strings.TrimSpace(os.Getenv(key)), "/"); v != "" {
			return v
		}
	}
	v := strings.TrimRight(strings.TrimSpace(configured), "/")
	if v == "" {
		return DefaultHost()
	}
	return v
}

func ValidModelName(s string) bool {
	s = strings.TrimSpace(s)
	if s == "" || len(s) > 128 {
		return false
	}
	if strings.HasPrefix(s, ".") || strings.HasPrefix(s, "-") || strings.HasPrefix(s, "/") || strings.HasPrefix(s, ":") {
		return false
	}
	if strings.HasSuffix(s, ".") || strings.HasSuffix(s, "-") || strings.HasSuffix(s, "/") || strings.HasSuffix(s, ":") {
		return false
	}
	if strings.Contains(s, "..") || strings.Contains(s, "//") || strings.Contains(s, " ") {
		return false
	}
	for _, c := range s {
		if (c >= 'a' && c <= 'z') || (c >= 'A' && c <= 'Z') || (c >= '0' && c <= '9') {
			continue
		}
		switch c {
		case '.', '_', '-', ':', '/':
		default:
			return false
		}
	}
	return true
}

func DefaultTemperature() float32 { return 0.8 }

func ClampTemperature(t float32) float32 {
	if t != t { // NaN
		return 0.8
	}
	if t < 0 {
		return 0
	}
	if t > 2 {
		return 2
	}
	return t
}

func CleanReply(text string) string {
	var sb strings.Builder
	blanks := 0
	for _, line := range strings.Split(text, "\n") {
		if strings.TrimSpace(line) == "" {
			blanks++
			if blanks <= 1 {
				sb.WriteString("\n")
			}
		} else {
			blanks = 0
			sb.WriteString(line)
			sb.WriteString("\n")
		}
	}
	return strings.TrimSpace(sb.String())
}

func StripLeadingSpeaker(text string, names []string) string {
	out := strings.TrimLeft(text, " \t\n\r")
	for {
		t := strings.TrimLeft(out, " \t\n\r")
		if !strings.HasPrefix(t, "[") {
			break
		}
		end := strings.Index(t, "]")
		if end <= 1 || end > 65 {
			break
		}
		after := strings.TrimLeft(t[end+1:], " \t\n\r")
		if !strings.HasPrefix(after, ":") {
			break
		}
		out = strings.TrimLeft(after[1:], " \t\n\r")
	}
	ordered := append([]string(nil), names...)
	// longest first
	for i := 0; i < len(ordered); i++ {
		for j := i + 1; j < len(ordered); j++ {
			if len(ordered[j]) > len(ordered[i]) {
				ordered[i], ordered[j] = ordered[j], ordered[i]
			}
		}
	}
	for {
		t := strings.TrimLeft(out, " \t\n\r")
		hit := false
		for _, n := range ordered {
			n = strings.TrimSpace(n)
			if n == "" {
				continue
			}
			if len(t) > len(n) && t[:len(n)] == n && strings.HasPrefix(t[len(n):], ":") {
				out = strings.TrimLeft(t[len(n)+1:], " \t\n\r")
				hit = true
				break
			}
		}
		if !hit {
			break
		}
	}
	return out
}

var metaStarters = []string{
	"let me summarize",
	"just to sum it up",
	"just to summarize",
	"let's simplify",
	"let me simplify",
	"to simplify,",
	"too much info",
	"i got carried away",
	"i think i got",
	"here is a summary",
	"here's a summary",
	"to summarize,",
	"in summary,",
}

func metaPreambleLen(text string) int {
	t := strings.TrimLeft(text, " \t\n\r")
	low := strings.ToLower(t)
	for _, s := range metaStarters {
		if !strings.HasPrefix(low, s) {
			continue
		}
		rest := t[len(s):]
		if strings.HasSuffix(s, ",") || strings.HasPrefix(rest, ":") || strings.HasPrefix(rest, ",") {
			cut := len(s)
			if strings.HasPrefix(rest, ":") || strings.HasPrefix(rest, ",") {
				cut = len(s) + 1
			}
			if strings.TrimSpace(t[cut:]) == "" {
				return 0
			}
			return cut
		}
		idx := strings.IndexAny(rest, ".!?")
		if idx < 0 {
			return 0
		}
		cut := len(s) + idx + 1
		if strings.TrimSpace(t[cut:]) == "" {
			return 0
		}
		return cut
	}
	return 0
}

func StripMetaPreamble(text string) string {
	out := text
	for {
		cut := metaPreambleLen(out)
		if cut == 0 {
			break
		}
		out = strings.TrimLeft(strings.TrimLeft(out, " \t\n\r")[cut:], " \t\n\r")
	}
	return out
}

func SanitizeReply(raw string, names []string) string {
	return StripMetaPreamble(StripLeadingSpeaker(CleanReply(raw), names))
}

func FinalizeReply(channel uint64, tagged string, names []string, raw string) (string, error) {
	text := SanitizeReply(raw, names)
	if strings.TrimSpace(text) == "" {
		return "", fmt.Errorf("ollama returned an empty reply")
	}
	push(channel, "user", tagged)
	push(channel, "assistant", text)
	return text, nil
}

func IsRateLimitErr(s string) bool {
	t := strings.ToLower(s)
	return strings.Contains(t, "429") ||
		strings.Contains(t, "rate limit") ||
		strings.Contains(t, "rate_limit") ||
		strings.Contains(t, "rate-limited") ||
		strings.Contains(t, "too many requests") ||
		strings.Contains(t, "quota")
}

func IsAPIFullErr(s string) bool {
	if IsRateLimitErr(s) {
		return true
	}
	t := strings.ToLower(s)
	return strings.Contains(t, "503") ||
		strings.Contains(t, "529") ||
		strings.Contains(t, "overload") ||
		strings.Contains(t, "capacity") ||
		strings.Contains(t, "server is busy") ||
		strings.Contains(t, "try again in a bit") ||
		strings.Contains(t, "api full")
}

func APIFullMessage() string {
	return "Sorry, I'm running hot right now (API full/rate limited) — try again in a minute."
}

func BuildTranscript(past []HistoryItem, tagged string, names []string) string {
	var sb strings.Builder
	for _, e := range past {
		if e.Role == "assistant" {
			cleaned := SanitizeReply(e.Content, names)
			if strings.TrimSpace(cleaned) == "" || StaleHistoryLine(cleaned) {
				continue
			}
			sb.WriteString(cleaned)
		} else {
			sb.WriteString(e.Content)
		}
		sb.WriteString("\n")
	}
	sb.WriteString(tagged)
	return sb.String()
}

type generateResponse struct {
	Response *string `json:"response"`
	Message  *struct {
		Content string `json:"content"`
	} `json:"message"`
}

func generateOnce(url, model, system, prompt string, temperature float32, think bool) (string, error) {
	req := map[string]interface{}{
		"model":      model,
		"prompt":     prompt,
		"stream":     false,
		"think":      think,
		"keep_alive": "10m",
		"options":    map[string]interface{}{"temperature": ClampTemperature(temperature)},
	}
	if strings.TrimSpace(system) != "" {
		req["system"] = system
	}
	body, _ := json.Marshal(req)
	resp, err := httpClient.Post(url, "application/json", bytes.NewReader(body))
	if err != nil {
		return "", err
	}
	defer resp.Body.Close()
	raw, _ := io.ReadAll(io.LimitReader(resp.Body, 1<<20))
	if resp.StatusCode < 200 || resp.StatusCode >= 300 {
		s := strings.TrimSpace(string(raw))
		if len(s) > 300 {
			s = string([]rune(s)[:300])
		}
		return "", fmt.Errorf("ollama %s: %s", resp.Status, s)
	}
	var parsed generateResponse
	if err := json.Unmarshal(raw, &parsed); err != nil {
		return "", err
	}
	if parsed.Response != nil && strings.TrimSpace(*parsed.Response) != "" {
		return *parsed.Response, nil
	}
	if parsed.Message != nil && strings.TrimSpace(parsed.Message.Content) != "" {
		return parsed.Message.Content, nil
	}
	return "", fmt.Errorf("ollama returned an empty reply")
}

func StaleHistoryLine(s string) bool {
	return strings.Contains(strings.ToLower(strings.TrimSpace(s)), "running hot right now")
}

func GlitchText(host, model, systemPrompt string, temperature float32, think bool) string {
	url := strings.TrimRight(host, "/") + "/api/generate"
	raw, err := generateOnce(url, model, systemPrompt, "You just glitched out. Tell the user in one short sentence, no details.", temperature, think)
	if err != nil {
		if IsAPIFullErr(err.Error()) {
			return APIFullMessage()
		}
		return "sorry, glitched out — try again in a sec"
	}
	text := StripMetaPreamble(StripLeadingSpeaker(CleanReply(raw), nil))
	if strings.TrimSpace(text) == "" {
		return "sorry, glitched out — try again in a sec"
	}
	if IsAPIFullErr(text) {
		return APIFullMessage()
	}
	r := []rune(text)
	if len(r) > 300 {
		return string(r[:300])
	}
	return text
}

func OllamaChat(host, model string, channel uint64, speaker, prompt, systemPrompt string, temperature float32, think bool) (string, error) {
	host = strings.TrimRight(host, "/")
	url := host + "/api/generate"
	tagged := fmt.Sprintf("[%s]: %s", SpeakerTag(speaker), prompt)
	past := snapshot(channel)
	fmt.Fprintf(os.Stderr, "ai chat: model=%s sys_chars=%d hist_msgs=%d temp=%v\n", model, len([]rune(systemPrompt)), len(past), ClampTemperature(temperature))
	names := []string{SpeakerTag(speaker)}
	for _, e := range past {
		if e.Role == "assistant" {
			continue
		}
		t := strings.TrimLeft(e.Content, " \t")
		if strings.HasPrefix(t, "[") {
			if end := strings.Index(t, "]"); end > 1 && end <= 64 {
				n := strings.TrimSpace(t[1:end])
				found := false
				for _, x := range names {
					if x == n {
						found = true
						break
					}
				}
				if n != "" && !found {
					names = append(names, n)
				}
			}
		}
	}
	transcript := BuildTranscript(past, tagged, names)
	first, err := generateOnce(url, model, systemPrompt, transcript, temperature, think)
	if err != nil {
		if IsAPIFullErr(err.Error()) {
			fmt.Fprintln(os.Stderr, "ollama_chat: api full on first call, offline fallback")
			fb := APIFullMessage()
			push(channel, "user", tagged)
			push(channel, "assistant", fb)
			return fb, nil
		}
		return "", err
	}
	return FinalizeReply(channel, tagged, names, first)
}

func ClearHistory(channel uint64) {
	histMu.Lock()
	defer histMu.Unlock()
	delete(hist, channel)
}

func ClearAllHistory() {
	histMu.Lock()
	defer histMu.Unlock()
	hist = map[uint64]*list.List{}
}

func RecordArtixy(channel uint64, text string) {
	t := strings.TrimSpace(text)
	if t == "" {
		return
	}
	r := []rune(t)
	if len(r) > 1500 {
		t = string(r[:1500])
	}
	push(channel, "assistant", t)
}

func ModelPresent(host, model string) *bool {
	url := strings.TrimRight(host, "/") + "/api/tags"
	resp, err := httpClient.Get(url)
	if err != nil {
		return nil
	}
	defer resp.Body.Close()
	if resp.StatusCode < 200 || resp.StatusCode >= 300 {
		return nil
	}
	var tags struct {
		Models []struct {
			Name string `json:"name"`
		} `json:"models"`
	}
	if err := json.NewDecoder(resp.Body).Decode(&tags); err != nil {
		return nil
	}
	want := strings.ToLower(strings.TrimSpace(model))
	base := strings.SplitN(want, ":", 2)[0]
	for _, m := range tags.Models {
		n := strings.ToLower(m.Name)
		nbase := strings.SplitN(n, ":", 2)[0]
		if n == want || nbase == base {
			b := true
			return &b
		}
	}
	b := false
	return &b
}

func SpeakerTag(raw string) string {
	flat := strings.Join(strings.Fields(raw), " ")
	var sb strings.Builder
	for _, c := range flat {
		switch c {
		case '[', ']':
			sb.WriteRune(' ')
		default:
			if c < 32 || c == 127 {
				sb.WriteRune(' ')
			} else {
				sb.WriteRune(c)
			}
		}
	}
	clean := strings.Join(strings.Fields(sb.String()), " ")
	r := []rune(clean)
	if len(r) > 64 {
		clean = string(r[:64])
	}
	if strings.TrimSpace(clean) == "" {
		return "someone"
	}
	return clean
}

func StripMention(content string, botID uint64) string {
	s := strings.ReplaceAll(content, fmt.Sprintf("<@%d>", botID), "")
	s = strings.ReplaceAll(s, fmt.Sprintf("<@!%d>", botID), "")
	return strings.TrimSpace(s)
}

func MentionsName(content string) bool {
	return strings.Contains(strings.ToLower(content), "artixy")
}

func StripName(content string) string {
	var out []string
	for _, w := range strings.Fields(content) {
		if !strings.Contains(strings.ToLower(w), "artixy") {
			out = append(out, w)
		}
	}
	return strings.Join(out, " ")
}

func ChunkReply(s string) []string {
	const max = 1900
	const maxChunks = 4
	s = strings.TrimSpace(s)
	if len([]rune(s)) <= max {
		return []string{s}
	}
	var chunks []string
	var cur strings.Builder
	curLen := 0
	flush := func() {
		t := strings.TrimRight(cur.String(), "\n")
		if strings.TrimSpace(t) != "" {
			chunks = append(chunks, t)
		}
		cur.Reset()
		curLen = 0
	}
	for _, line := range strings.Split(s, "\n") {
		lineLen := len([]rune(line)) + 1
		if lineLen > max {
			if strings.TrimSpace(cur.String()) != "" {
				flush()
			}
			r := []rune(line)
			for i := 0; i < len(r); i += max {
				end := i + max
				if end > len(r) {
					end = len(r)
				}
				chunks = append(chunks, string(r[i:end]))
				if len(chunks) >= maxChunks {
					break
				}
			}
			if len(chunks) >= maxChunks {
				break
			}
			continue
		}
		if curLen+lineLen > max {
			flush()
			if len(chunks) >= maxChunks {
				break
			}
		}
		cur.WriteString(line)
		cur.WriteString("\n")
		curLen += lineLen
	}
	if strings.TrimSpace(cur.String()) != "" && len(chunks) < maxChunks {
		flush()
	}
	if len(chunks) == 0 {
		r := []rune(s)
		if len(r) > max {
			r = r[:max]
		}
		return []string{string(r)}
	}
	return chunks
}
